### ISSUE TITLE
[Feature Request]: Automatic Preset Assigning

### ISSUE THREAD

**Issue by DerIshmaelite** [open-spaced-repetition/fsrs4anki#709]
The number of cards in a deck sharing a single preset has a massive influence on RMSE based on personal experience. But having too many cards with varying intrinsic complexities (and difficulty) could decrease the accuracy. Therefore, I think there has to be some theoretical "midpoint" where the **number of cards per preset** is optimized and the best possible RMSE is achieved.

**Comment by aleksejrs** [comment]
I found that at least some presets with similar minimum recommended retention could be merged without harming RMSE much, and those with different MRR not.

**Comment by Xemorr** [comment]
I think this closed issue by me years ago is related https://github.com/open-spaced-repetition/fsrs4anki/issues/616, or suggests how to go about automatic preset assigning imo

**Comment by Expertium** [comment]
You're not gonna believe this - I literally just finished 46 experiments related to smart preset assignment, and 0 of them outperformed the baseline.
You can find the code, .jsonl result files and the .xlsx summary file here: https://github.com/open-spaced-repetition/srs-benchmark-rust

**Comment by Xemorr** [comment]
> You're not gonna believe this - I literally just finished 46 experiments related to smart preset assignment, and 0 of them outperformed the baseline. You can find the code, .jsonl result files and the .xlsx summary file here: https://github.com/open-spaced-repetition/srs-benchmark-rust

What is your explanation for the [lower RMSE I witnessed](https://github.com/open-spaced-repetition/fsrs4anki/issues/616#issuecomment-1970805789) when separating out cloze and basic cards for example

I suppose it is old, so I'd need to reprove it worked on those decks, and I've finished studying for university so that's quite difficult 😅 

From my interpretation of your commit message, you're clustering at the _deck_ level. The original feature request here talks about 
> The number of cards in a deck sharing a single preset has a massive influence on RMSE

I would argue performing clustering at the deck level is not sufficient to disprove OP as decks are usually split out by _topic_ not by anything to do with layout of card normally.

**Comment by Expertium** [comment]
I'll try some different approaches, but for now smart presets are dead until proven otherwise

**Comment by Expertium** [comment]
Yesterday I did 46 experiments with a distance metric that measured similarity between *FSRS parameters* and 46 different variations of clustering algorithms. None beat the "same parameters for every deck" baseline. Today I tried a different distance metric that measures similarity between *predictions*, aka instead of asking "Are parameters similar?", we ask "Are predicted probabilities similar?". Only 1/46 beat the baseline, and by a negligible, within-noise amount.
Now it's time to abandon distance metrics and just do things the hard way: tons and tons of optimizations. Instead of measuring deck similarity, we abandon that idea entirely and just measure whether lumping decks together helps to reduce log loss, regardless of their "similarity" in any sense (well, unless the user has tons of decks, in which case we have to use distance metrics to lump them together to bring the number of decks down to a reasonable number, then we do optimizations). If even that doesn't help, then this idea is truly dead.

EDIT: nope, the final approach didn't help either. Smart preset assignment is truly dead.

**Comment by L-M-Sherlock** [comment]
@Expertium do you use different parameters for different presets? Maybe the dataset of the benchmark is just shit😂

**Comment by Expertium** [comment]
https://github.com/open-spaced-repetition/srs-benchmark-rust
You can check the code here. Basically, I used decks as smallest units, optimized parameters for each deck (like `--partitions deck` does), and then ran several clustering algorithms that put decks together into clusters (pseudo-presets, whatever you wanna call them), optimize parameters of the whole cluster together, and then those parameters are evaluated in the test fold.
Or, did you mean whether I *personally* use different params for different presets? Yeah, I do. Placebo 😅

**Comment by L-M-Sherlock** [comment]
> Or, did you mean whether I _personally_ use different params for different presets? Yeah, I do. Placebo 😅

Yeah, I mean that. Why did you say that it's Placebo for yourself? Did you try those approaches on your collection?

**Comment by Expertium** [comment]
Nope. I'm saying it's placebo because so far no approach has yielded good results on average, across 10k users, so it's extremely unlikely that I would get any measurable benefit.

**Comment by L-M-Sherlock** [comment]
MOVING-AVG is very good across 10k users, but its performance is bad on my collection. That's why I have questions about the dataset. I think it's still a good trial to test it on your collection. Maybe there is misalignment between the benchmark and real collection.
